### PR TITLE
[Mono.Security] Do not decode data beyond detected length in ASN1 parser

### PR DESCRIPTION
**Pull request by SergiusTheBest**
If `byte[] data` has additional data at the end `Decode` will try to parse it and will generate incorrect tags or throw an exception. This fix makes `Decode` to use only `nLength` bytes from `byte[] data`.

### CODE REVIEW

File: mcs/class/Mono.Security/Mono.Security/ASN1.cs [hunk 1]
```diff
@@ -86,8 +86,8 @@ public ASN1 (byte[] data)
 			Buffer.BlockCopy (data, (2 + nLenLength), m_aValue, 0, nLength);
 
 			if ((m_nTag & 0x20) == 0x20) {
-				int nStart = (2 + nLenLength);
-				Decode (data, ref nStart, data.Length);
+				int nStart = 0;
+				Decode(m_aValue, ref nStart, m_aValue.Length);
```
Comment (filipnavara) [inline mcs/class/Mono.Security/Mono.Security/ASN1.cs]: nit: Mono code style should be `Decode (` (with space)

**Comment by filipnavara** [comment]
@SergiusTheBest Btw, not sure how you managed to hit it but using Mono.Security for anything serious is highly discouraged. If you need ASN.1 parser in your project then there's one in https://github.com/dotnet/corefxlab with NuGet published at https://dotnet.myget.org/feed/dotnet-corefxlab/package/nuget/System.Security.Cryptography.Asn1.Experimental.

**Comment by lambdageek** [comment]
@monojenkins build

**Comment by SergiusTheBest** [comment]
@filipnavara I'm using it to parse digital signatures from PE files (not directly `ASN1` but more high-level classes that uses `ASN1` internally).

File: mcs/class/Mono.Security/Mono.Security/ASN1.cs (hunk 1 above)
Comment (SergiusTheBest) [reply mcs/class/Mono.Security/Mono.Security/ASN1.cs]: fixed

**Comment by filipnavara** [comment]
You may be able to access Authenticode signatures with public APIs. On some platforms it's possible to use `X509Certificate.CreateFromSignedFile` or alternative high-level APIs ([1](https://github.com/dotnet/corefx/issues/2910#issuecomment-374787612), [2](https://github.com/dotnet/corefx/issues/2910#issuecomment-399385948)).

**Comment by SergiusTheBest** [comment]
@filipnavara Unfortunately they don't support multiple signatures and return only a primary one.

**Comment by filipnavara** [comment]
I believe the approach with extracting the PE block and then using `SignedCms` to decode it (links above) is able to use more signatures. I can give it a try later.

**Comment by SergiusTheBest** [comment]
@filipnavara I just tested `SignedCms`. It successfully decoded a signature but only a primary one:
```csharp
var cms = new SignedCms();
cms.Decode(data); // data contains security directory from PE file
var count = cms.Certificates.Count; // here is a collection with only 1 certificate
```

**Comment by filipnavara** [comment]
I vaguely remember that the format to store multiple signatures is actually several `WIN_CERTIFICATE` structures (https://docs.microsoft.com/en-us/windows/win32/api/wintrust/ns-wintrust-win_certificate) listed one after another in the PE `IMAGE_DIRECTORY_ENTRY_SECURITY` section. For each of these `SignedCms` would return only one certificate but there could be multiple of them in the same section.

**Comment by SergiusTheBest** [comment]
@filipnavara No, there is only one `WIN_CERTIFICATE` and it contains a primary signature. All other signature are stored as attributes with OID `1.3.6.1.4.1.311.2.4.1` in the primary signature.

**Comment by SergiusTheBest** [comment]
@filipnavara However it's possible to check those attributes and parse them using `SignedCms`: https://www.sysadmins.lv/blog-en/reading-multiple-signatures-from-signed-file-with-powershell.aspx

**Comment by filipnavara** [comment]
I stand corrected on the `WIN_CERTIFICATE` layout. I tried to check it on one of our test files and it had only single one of those.

Anyway, I recommend using `SignedCms` if at all possible. The implementation in Mono is taken from CoreFX (.NET Core) and uses basically the same ASN.1 parser I linked above. It was actually reviewed by security experts unlike the old Mono.Security code.

**Comment by SergiusTheBest** [comment]
@filipnavara Yes, thank you for pointing me into `SignedCms`.

**Comment by akoeplinger** [comment]
@monojenkins build

**Pull request merged**